### PR TITLE
System Functions brush-up

### PR DESCRIPTION
**Pull request by Claes65**

### CODE REVIEW

**Comment by Claes65** [comment]
If you can show me how to do a type check, then I can add that as well before we close this @MarkusAmshove

**Comment by Claes65** [comment]
pls check against your source code, @MarkusAmshove - I am happy from my side now.

File: libs/natparse/src/main/java/org/amshove/natparse/natural/IAtnOperandNode.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package org.amshove.natparse.natural;
+
+public interface IAtnOperandNode extends IOperandNode
```
Comment (MarkusAmshove) [inline libs/natparse/src/main/java/org/amshove/natparse/natural/IAtnOperandNode.java:3]: To simplify the code, lets create an interface `IMathFunctionOperandNode` that extends `IOperandNode` and make every one of these interfaces implement `IMathFunctionOperandNode`

If all these functions only have one parameter, the interface could be 

```java
public interface IMathFunctionOperandNode extend IOperandNode
{
    IOperandNode parameter();
}
```

Then we can make all interfaces like `IAtnOperandNode`, `IAverOperandNode` etc. empty.

File: libs/natparse/src/main/java/org/amshove/natparse/parsing/TypeChecker.java [hunk 2]
```diff
@@ -193,17 +193,148 @@ private void checkNode(ISyntaxNode node)
...
+			return;
 		}
 
 		if (node instanceof ISystemFunctionNode sysFuncNode)
 		{
 			checkSystemFunctionParameter(sysFuncNode);
+			return;
+		}
+
+		if (checkMathematicalSystemFunctions(node))
```
Comment (MarkusAmshove) [inline libs/natparse/src/main/java/org/amshove/natparse/parsing/TypeChecker.java:211]: Then change this to 
```java
if (node instanceof IMathFunctionOperandNode function)
{
    checkMathematicalSystemFunction(function);
    return;
}
```

File: libs/natparse/src/main/java/org/amshove/natparse/parsing/TypeChecker.java [hunk 3]
```diff
@@ -193,17 +193,148 @@ private void checkNode(ISyntaxNode node)
...
+		{
+			return;
+		}
+
+		checkAlphaSystemFunctions(node);
+	}
+
+	private boolean checkMathematicalSystemFunctions(ISyntaxNode node)
```
Comment (MarkusAmshove) [inline libs/natparse/src/main/java/org/amshove/natparse/parsing/TypeChecker.java:219]: And this

```java
private void checkMathematicalSystemFunction(IMathSystemFunctionOperandNode operand)
{
   var type = inferDataType(operand.parameter());
   if (!type.isNumericFamily())
    {
        report(...);
    }
}

File: libs/natparse/src/main/java/org/amshove/natparse/natural/IAtnOperandNode.java (hunk 1 above)
Comment (MarkusAmshove) [reply libs/natparse/src/main/java/org/amshove/natparse/natural/IAtnOperandNode.java:3]: If I miss some functions that take more than one parameter my approach won't work :-)

**Review by MarkusAmshove — CHANGES_REQUESTED** [review]
We can simplify the code a bit by introducing a new interface for these type of functions. If some functions take multiple parameter, we might have to add different interfaces.

Also I get around 28k new diagnostics, but those are all around copy codes. I'll try to see if I find a quick solution so that we don't need to have #125 in place before we merge this

**Comment by Claes65** [comment]
> We can simplify the code a bit by introducing a new interface for these type of functions. If some functions take multiple parameter, we might have to add different interfaces.
> 
> Also I get around 28k new diagnostics, but those are all around copy codes. I'll try to see if I find a quick solution so that we don't need to have #125 in place before we merge this

Excellent! Thanks for this. I’ll do the changes tomorrow as I’m ooo today. I hope you find a solution to cc - curious to what the issue is.

File: libs/natparse/src/main/java/org/amshove/natparse/natural/IAtnOperandNode.java (hunk 1 above)
Comment (Claes65) [reply libs/natparse/src/main/java/org/amshove/natparse/natural/IAtnOperandNode.java:3]: > If I miss some functions that take more than one parameter my approach won't work :-)

I’m quite sure the only take one. :)

**Review by MarkusAmshove — APPROVED** [review]

**Comment by MarkusAmshove** [comment]
I get 11 new diagnostics, but they'll be gone when type information are loaded for DDM fields

**Pull request merged**